adding section 5 on legacy package management (#2123)
Azure/azure-sdk: docs/policies/releases.md

### 5) Updating package manager references

Follow the best practices below to ensure that the visibility of the new release is elevated.

Identify supported legacy libraries for the SDK you are releasing as early as possible <br>

> -	Note, we cannot deprecate GA libraries which were released less than 1 year ago per the [current Azure SLA](https://www.microsoft.com/EN-US/LICENSING/PRODUCT-LICENSING/PRODUCTS) 

Reviewer: ```suggestion
-	Note, we cannot deprecate GA libraries which were released less than 1 year ago per the [current Azure SLA](https://www.microsoft.com/LICENSING/PRODUCT-LICENSING/PRODUCTS) 
```

- Note,  We should direct users to the new library only if it is covering the legacy library specs. Make sure you confirm feature and performance parity at your arch board reviews.

> -	Add all identified packages to the [Package Deprecation Plan](https://microsoft-my.sharepoint.com/:w:/p/elraikhm/ETPskYQ1MGxErAdJf5u0OaYBJjp6j1xBBQwuU74jmTx5xw?e=4b3oyj)

Reviewer: Is this a publicly accessible link? If not we should fix that as we want these guidelines to be all in the public.

Author: @salameer,  I was under impression that this policy is internal only.

Reviewer 2: Hi @elraikhm, I would think dependency info would be relevant for external users.

Author: @salameer , I'm confused, there is no dependency info in this file :) this file is to track our progress as we are deprecating libraries, there is an external link where we post similar info externally  https://azure.github.io/azure-sdk/releases/latest/. 
We are waiting for Azure Global to finalize the overall azure deprecation and retirement policy with legal to make our deprecation strategy public

For each legacy package in your list

> -	Update the legacy package README with a link to the new package.

Reviewer: Where is the legacy README? Should we provide a link?

```markdown
Please note, a newer package is available [package name](https://dummylinktopackage.com/) as of [date-MM/YY]. We strongly encourage you to upgrade. 
See [Migration Guide](https://dummylink.com/) for more details.
```

> -	Update the legacy Samples README with link to the new samples.

Reviewer: I suspect it will be very difficult to find all the legacy samples and various readmes to actually be able to update these.

Author: Updated to point to the docs samples root

Reviewer 3: There is ongoing effort in identifying all samples and docs under docs.microsoft.com that refer to older packages. We should be seeing progress in the new semester. So, this should be doable

```markdown
Samples in this repository use older packages [list legacy package used]. We recommend that you get started using the new package here instead: <link>.
```
-	Update the new package README with a link to the migration guides.
```markdown
This document covers [library name] [package name](https://dummylinktopackage.com/). 
If you're using an older package, we recommend that you upgrade your code.
See [Migration Guide](https://dummylink.com/) for more details
```
-	Update package index files to hide legacy packages and identify legacy packages replacement
```markdown

> Go to https://github.com/Azure/azure-sdk/blob/master/_data/releases/latest/ and update [language]-packages.csv.

Reviewer: cc @Petermarcu @jongio I don't think we want to hide legacy versions we just want to add a note to the old one pointing to the new one and the new one pointing to the old one.

Reviewer 4: That is what we've been doing yes.

> Use Hide collumn to hide deprecated packages, use Notes collumn to clarify what is being replaced by the new package

Reviewer: ```suggestion
Use Hide collumn to hide deprecated packages, use Notes column to clarify what is being replaced by the new package
```

### 5) Updating package manager references

Follow the best practices below to ensure that the visibility of the new release is elevated.

Identify supported legacy libraries for the SDK you are releasing as early as possible <br>

> -	Note, we cannot deprecate GA libraries which were released less than 1 year ago per the [current Azure SLA](https://www.microsoft.com/LICENSING/PRODUCT-LICENSING/PRODUCTS) 

Author: Wait for one year since the release

> - Note,  We should direct users to the new library only if it is covering the legacy library specs. Make sure you confirm feature and performance parity at your arch board reviews.

Author: Agree, suggest to add this to the migration guides guidance section (@sameralameer )

-	Add all identified packages to the Package Deprecation Plan.
For example,  https://github.com/Azure/azure-storage-net which would be one of the legacy packages for azure storage sdk. 

> For each legacy package in your Package Deprecation Plan

Reviewer 3: There is no mention of any Quick Starts or Tutorials needing to be updated. Can we add that as well?

-	Update the legacy library package README with a link to the new package.
```markdown
Please note, a newer package is available [package name](https://dummylinktopackage.com/) as of [date-MM/YY]. We strongly encourage you to upgrade. 

> See [Migration Guide](https://dummylink.com/) for more details.

Reviewer 3: ```suggestion
See [Migration Guide](https://dummylink.com/) for benefits from upgrading and code snippets comparing the usage between the two packages.
```

```
-	Update the legacy Samples README with link to the new samples.

> Note, you can search for sample repositories under https://docs.microsoft.com/en-us/samples/browse/?products=azure 

Reviewer 3: Not all samples repos will be found when using the above link. We have seen many showing up looking directly under the Azure-Samples repo. For example: https://github.com/Azure-samples?q=%22event+hubs%22&type=&language=

Author: Is there are a reason why those samples are not discoverable from docs@microsoft? Should we add them there?

Reviewer 3: The readme in the samples repos need to have particular metadata for them to appear in docs.microsoft.com/samples/browse. Ideally they should, and we can follow up with the content team to see what they are doing to ensure this happens.

```markdown

> Samples in this repository use older packages [list legacy package here]. We recommend that you get started using the new package here instead: <link>.

Reviewer 3: Sometimes, the sample in the repo might be show casing multiple services or the use of both data and mgmt plane pkgs for the same service. In such cases, simply saying that one of the packages being used is legacy and pointing to another package would not be very user friendly. We would basically be saying "use this sample, and replace this particular package usage with another, and here is the migration guide for it". 

We need to have criteria to decided between
- updating the sample repo in place to use the new package
- archiving repo, and pointing people to the samples in our repo to use the new package
- have separate folders for new and old packages

Author: My thought process was: retain legacy samples and update their readme to point to the new samples in a separate repository.  I will update the wording as it appeared not clear.

Reviewer 3: That would require us to make copies of the same samples in separate repos with similar names. That can hurt SEO

```
-	Update the new package README with a link to the migration guides.
```markdown
This document covers [library name] [package name](https://dummylinktopackage.com/). 

> If you're using an older package, we recommend that you upgrade your code.

Reviewer 3: There can be more than one old package. The migration guide will not be covering all of them. So mentioning the old package name here is important.

Also, in case of JS and Python, we re-use the package name and simply release a new version. In that case this message should be different

Author: will update to include the version range.

See [Migration Guide](https://dummylink.com/) for more details
```
-	Update package index files to identify legacy packages replacement
```markdown
Go to https://github.com/Azure/azure-sdk/blob/master/_data/releases/latest/ and update [language]-packages.csv.
Use Notes column to clarify what is being replaced by the new package
```

> -	Make updates to the legacy packages in the package manager.

Reviewer 3: This section does not apply if we are re-using the package name and only releasing new versions like we do quite a bit in case of Python and JS. It is worth calling that out here.

Author: Agreed, will make a note

Use the following message and follow programming language instructions
```markdown
Please note, a newer package  is available [package name](https://dummylinktopackage.com/) as of [date-MM/YY].

> We strongly encourage you to upgrade. See [Migration Guide](https://dummylink.com/) for more details.

Reviewer 3: Not all languages support the use of markdown in the deprecation message. We can skip the link for the package and make do with just the package name. In case of migration guide, I would recommend we have an aka.ms link so that we can use a short url instead of the very long one pointing to the guide that lives deep inside our repo

Author: The link to the package in the package manager should not be too long. Do you know which languages do not support markdown?
https://www.npmjs.com/package/@azure/event-hubs, 
https://pypi.org/project/azure-eventhub/ 
https://www.nuget.org/packages/Azure.Messaging.EventHubs/
https://mvnrepository.com/artifact/com.microsoft.azure/azure-eventhubs

Reviewer 3: I was referring to the links to migration guides, not the package in package manager. We can add a note here to use the aka.ms links to ensure the links are short

```
Language|Instructions
----------|---------
.Net|Follow instructions at [deprecating NuGet packages](https://docs.microsoft.com/en-us/nuget/nuget-org/deprecate-packages) to deprecate the package with the message above.<br>Select all versions of the package you are deprecating.<br>Choose deprecation reason as "Other" because "Legacy" is for packages that are no longer maintained.
Java | Update the project description in the POM file to include the deprecation message. Publish Update.
TS/JS | Run the command `npm deprecate` which takes the package name and the deprecation message.

> Python | Publish an update to the package after updating the README.

Reviewer 5: Note that PyPI does a note that there is a newer version automatically:
https://pypi.org/project/azure-servicebus/0.50.3/

See on this screenshot, a track1 package on PyPI today
![image](https://user-images.githubusercontent.com/1050156/103375197-65e12f00-4a8e-11eb-9f3a-a4450f56da89.png)

So by default, I don't encourage re-release to PyPI to write something that PyPI is doing already.

However, if we renamed the package (e.g. textanalytics) we don't get this benefits and I agree an update is needed. Example:
https://pypi.org/project/azure-cognitiveservices-formrecognizer/

Author: We wanted to update the readme to point to the latest version from the readme text. Do you suggest just updating the readme, but do not re-release the package?

Reviewer 5: The Readme on master will be the latest version anyway, example with ServiceBus:
https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/servicebus/azure-servicebus/README.md

So I'm not sure what Readme you are referring to here?

To summarize:
- If a track2 package has kept track1 package name, nothing is necessary. We could do a disclaimer in the doc for the opposite actually ("Disclaimer: if you're here because you're looking for track1, this is where you get it")
- If a track2 package is a rename (azure-cognitiveservices-textanalytics => azure-ai-textanalytics), I agree with the propositions here (note to redirect, etc.)

@johanste for two cents if any

  |  

#### Legacy packages support after the update

-	Ensure you have the infrastructure in place to build and ship fixes for the legacy libraries for the next 3 years.<br> You will need to ship security and critical bug fixes for the first year.<br> You don’t have to proactively fix anything else, but a customer calling for support should be able to get it.
-	Encourage customers to migrate to the new library when possible from this point on.
-	After one year, customers will need to provide a strong business need to justify and receive legacy library updates. Most of the customers are expected to migrate to the latest library at this point.

> -	Monitor the deprecated library usage. If the number of customers is small, consider working with them individually to migrate to the latest version.<br> At the point when there are no customers left, you can retire the library following the [Azure Global Retirement Policy process](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37673/Retirements-and-Breaking-Changes).

Reviewer 3: > At the point when there are no customers left, you can retire the library

There will always be a long tail end. Are we really going to wait until absolutely no customers are left on the old version?

Author: No, we can reach out to customers and encourage/help migrate. We will not provide fixes/support unless there is a strong business justification.
